Add StarPrerouting to the Default Passmanagers

sbrandhsn opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Adds the star to linear prerouting pass to optimization level 2 and 3 passmanagers (see #11387)

### Details and comments

sbrandhsn committed:
add StarPrerouting

mtreinish commented on releasenotes/notes/adding-starprerouting-to-default-passmanager-0fb9c4a86ac2ad0d.yaml:
```diff
@@ -0,0 +1,7 @@
+    Added the :class:`StarPreRouting` pass to the default optimization level 2+3 passmanager. `StarPreRouting`
+    runs before a layout is fixed (i.e. on abstract circuits) and transforms a circuit with inner
+    'star'-connectivities to a linear connectivity before continuing to other transpilation stages, e.g.
+    the layouting and routing stage.
```
```suggestion
    the layout and routing stage.
```

mtreinish reviewed (dismissed):
This LGTM in it's current form (leaving an approval here just to indicate this now, but this can't merge until its dependencies do). There will be a merge conflict with #12111 but that should be simple to fix and I'm pretty sure we want to run this pass after `ElidePermutations` since this inserts permutations. :)

mtreinish commented on releasenotes/notes/adding-starprerouting-to-default-passmanager-0fb9c4a86ac2ad0d.yaml:
```diff
@@ -0,0 +1,7 @@
+---
+features:
```
```suggestion
features_transpiler:
```

sbrandhsn committed:
Update releasenotes/notes/adding-starprerouting-to-default-passmanager-0fb9c4a86ac2ad0d.yaml

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

sbrandhsn committed:
Update releasenotes/notes/adding-starprerouting-to-default-passmanager-0fb9c4a86ac2ad0d.yaml

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

sbrandhsn commented:
Sounds good, thanks! :-)

mtreinish committed:
Merge branch 'main' into add-star-prerouting

mtreinish committed:
Merge branch 'main' into add-star-prerouting

mtreinish committed:
Update preset passmanager construction and tests

The preset passmanager unit tests found a couple of edge cases that
challenged this PR's previous assumptions that we should run the pass
as part of the init stage prior to layout. The first issue flagged was
that we were previously running this pass without the presence of a
target/coupling map. This would potentially lead to unecessary swap
insertions as we are only pre-routing to potentially match a linear
connectivity on the target backend, or to make routing easier. We should
not be running StarPreRouting without a target with connectivity
constraints. The second issue identified by the tests was that in the
case when the input circuit has a perfect layout on the target and star
connectivity in the circuit the StarPreRouting pass would potentially
interfere with finding that perfect layout because it would insert swaps
which may break that perfect match. To correct for these issues instead
of running at the end of the init stage this commit migrates to using
the pass in the layout stage for optimziation levels 2 and 3. The new
workflow for the layout stage in these optimization levels is:

1. Try `VF2Layout`
2. If `VF2Layout` does **not** find a match run `StarPreRouting`
3. If `StarPreRouting` ran and found and replaced stars run `VF2Layout`
   again (to find a potential perfect linear layout)
4. If we do not have a layout from `VF2Layout` (either 1 or 3) run
   `SabreLayout`

Some other small tests are updated to account for changes in layout and
other transpiler outcomes as a function of running this pass now too.

sbrandhsn committed:
Update builtin_plugins.py

sbrandhsn committed:
making sure star pre-routing is not running twice

mtreinish committed:
Merge remote-tracking branch 'origin/main' into add-star-prerouting

mtreinish committed:
Fix lint

mtreinish committed:
Fix layout composition

mtreinish commented:
I ran some quick asv benchmarks with this, we might want to hold off on adding this to the preset passmanagers for now. It seems to have a pretty big negative impact on performance. It helps on QFT's quality significantly, but for the other circuits it has a big impact on runtime and/or output depth.

```
Benchmarks that have improved:

| Change   |   Before [43c065fc] <add-star-prerouting~3^2> |   After [46aa5cfb] <add-star-prerouting> |   Ratio | Benchmark (Parameter)                                       |
|----------|-----------------------------------------------|------------------------------------------|---------|-------------------------------------------------------------|
| -        |                                          2582 |                                     1540 |    0.6  | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cz')  |
| -        |                                          2582 |                                     1540 |    0.6  | utility_scale.UtilityScaleBenchmarks.track_qft_depth('ecr') |
| -        |                                          2582 |                                     1531 |    0.59 | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cx')  |

Benchmarks that have stayed the same:

| Change   | Before [43c065fc] <add-star-prerouting~3^2>   | After [46aa5cfb] <add-star-prerouting>   | Ratio   | Benchmark (Parameter)                                                         |
|----------|-----------------------------------------------|------------------------------------------|---------|-------------------------------------------------------------------------------|
|          | 2.42±0.01s                                    | 3.18±0.01s                               | ~1.31   | utility_scale.UtilityScaleBenchmarks.time_qaoa('ecr')                         |
|          | 21.5±0.03s                                    | 25.9±0.1s                                | ~1.20   | utility_scale.UtilityScaleBenchmarks.time_qft('cx')                           |
|          | 24.4±0.04s                                    | 26.2±0.1s                                | 1.07    | utility_scale.UtilityScaleBenchmarks.time_qft('cz')                           |
|          | 24.4±0.05s                                    | 25.4±0.1s                                | 1.04    | utility_scale.UtilityScaleBenchmarks.time_qft('ecr')                          |
|          | 9.11±0.05ms                                   | 9.12±0.04ms                              | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cz')               |
|          | 98.8±0.2ms                                    | 98.8±0.5ms                               | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cz')                |
|          | 99.1±0.3ms                                    | 99.2±0.7ms                               | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('ecr')               |
|          | 9.17±0.03ms                                   | 9.10±0.2ms                               | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cx')               |
|          | 9.17±0.04ms                                   | 9.09±0.03ms                              | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('ecr')              |
|          | 99.9±0.5ms                                    | 99.2±0.3ms                               | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cx')                |
|          | 32.3±0.07ms                                   | 32.1±0.1ms                               | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cx')  |
|          | 32.2±0.3ms                                    | 32.0±0.04ms                              | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cz')  |
|          | 32.5±0.2ms                                    | 32.2±0.1ms                               | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('ecr') |
|          | 444                                           | 414                                      | 0.93    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cx')      |
|          | 444                                           | 414                                      | 0.93    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cz')      |
|          | 444                                           | 414                                      | 0.93    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('ecr')     |

Benchmarks that have got worse:

| Change   | Before [43c065fc] <add-star-prerouting~3^2>   | After [46aa5cfb] <add-star-prerouting>   |   Ratio | Benchmark (Parameter)                                              |
|----------|-----------------------------------------------|------------------------------------------|---------|--------------------------------------------------------------------|
| +        | 1.28±0.02s                                    | 4.13±0.02s                               |    3.22 | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cx')  |
| +        | 1.78±0.01s                                    | 4.61±0.03s                               |    2.59 | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('ecr') |
| +        | 1.93±0.04s                                    | 4.80±0.01s                               |    2.49 | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cz')  |
| +        | 1.13±0.01s                                    | 1.87±0.01s                               |    1.66 | utility_scale.UtilityScaleBenchmarks.time_qaoa('cx')               |
| +        | 1607                                          | 2331                                     |    1.45 | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cx')        |
| +        | 1622                                          | 2331                                     |    1.44 | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cz')        |
| +        | 1622                                          | 2331                                     |    1.44 | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('ecr')       |
| +        | 2.92±0.02s                                    | 3.76±0.02s                               |    1.29 | utility_scale.UtilityScaleBenchmarks.time_qaoa('cz')               |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

alexanderivrii commented on qiskit/transpiler/passes/optimization/elide_permutations.py:
```diff
@@ -105,8 +105,10 @@ def _apply_mapping(qargs):
         new_layout = Layout({dag.qubits[out]: idx for idx, out in enumerate(qubit_mapping)})
-        if current_layout := self.property_set["virtual_permutation_layout"] is not None:
-            self.property_set["virtual_permutation_layout"] = current_layout.compose(new_layout)
+        if current_layout := self.property_set["virtual_permutation_layout"]:
+            self.property_set["virtual_permutation_layout"] = new_layout.compose(
+                current_layout.inverse(dag.qubits, dag.qubits), dag.qubits
+            )
```
Oh right, this seems to be the correct order for composing final permutations, thanks for fixing this! If our circuit is `C -- P`, with `P` implicitly tracked as a "final permutation", and then we further rewrite `C` as `D -- Q`, then the original circuit becomes `D -- Q -- P`, i.e. we now want to implicitly track the "final permutation" for `Q -- P`.

alexanderivrii replied:
This leads to a very interesting general question. If we consider SWAPs in the "final permutation" part of the circuit to be free (come at no cost), then what is the best way to "massage" a quantum circuit (e.g. inserting/deleting SWAPs and suitably changing the following gates) to make it easier to route.

alexanderivrii commented on qiskit/transpiler/preset_passmanagers/builtin_plugins.py:
```diff
@@ -778,6 +783,8 @@ def _swap_mapped(property_set):
                 skip_routing=pass_manager_config.routing_method is not None
                 and pass_manager_config.routing_method != "sabre",
             )
+            layout.append(ConditionalController(StarPreRouting(), condition=_vf2_match_not_found))
+            layout.append(ConditionalController(choose_layout_0, condition=_star_prerouting_check))
```
Yeah, very good point that `StarPreRouting` may in cases make the circuit harder to route (and in cases to destroy the perfect match).

alexanderivrii reviewed (approved):
I am approving this, though looking at the ASV results it is indeed not clear that we want the `StarPreRouting` pass to run by default. 

It does seem that a more thorough investigation of when and how to run the pass, e.g.  maybe the pass would be beneficial if we only rewrite the "large stars", or only for low-degree coupling maps. 

But in any case, it would be good to include the fix to composing virtual pre-layouts.

mtreinish commented:
Given the open performance questions on this PR I think it's prudent for us to just defer this until the next release. The depth improvements for qft and also the square heisenberg circuit look really nice. But the qaoa circuit is concerning enough that we shouldn't rush into doing this. I'll split out the layout composition fix into a separate PR that we can include in 1.1.0 and in parallel lets investigate adding it to the preset pass managers.

ShellyGarion commented:
> Given the open performance questions on this PR I think it's prudent for us to just defer this until the next release. The depth improvements for qft and also the square heisenberg circuit look really nice. But the qaoa circuit is concerning enough that we shouldn't rush into doing this. I'll split out the layout composition fix into a separate PR that we can include in 1.1.0 and in parallel lets investigate adding it to the preset pass managers.

I think that QAOA has a good swap strategies algorithm by https://arxiv.org/abs/2202.03459.
I think it was implemented in Qiskit as part of the evolution gates synthesis (but not in general).

mtreinish commented:
I'm going to close this PR because I don't think we're planning to do this any time soon. Linear routing a star interaction doesn't seem to be an improvement in all cases at least in the approach this and if we want to integrate this into the preset pass manager we'll need to do more discovery around how to best do this.
